rosbridge_server/src/rosbridge_server/websocket_handler.py

```diff
@@ -197,3 +219,121 @@ def get_compression_options(self):
+        auth_srv = rospy.ServiceProxy(self.auth_service_name, GetLiveViewAuth)
+
+        if "token" not in msg:
+            self.send_status("Message is malformed. Must include `token`.", "error")
+            return
+
+        rospy.loginfo(f"Calling {self.auth_service_name} to check auth for user...")
+        response = auth_srv(msg["token"])
+
+        # An internal error. Handle it locally and close the connection. This needs to be fixed, not handled.
+        if response.result == GetLiveViewAuthResponse.FAILURE:
+            reason = f"Could not auth user: {msg['username']}. Service failed with message: {response.message}"
+            self.send_status(reason, "error")
+            self.sendClose()
+            return
+
+        # A 403-like error. Tell the client of this failure and then close connection.
+        if response.result == GetLiveViewAuthResponse.INVALID_CREDENTIALS:
+            reason = f"Invalid credentials. Reason: {response.message}"
+            self.send_status(reason, "error")
+            return
+
+        # Auth worked. Set RosBridge state for authed/permissions, and send a response.
+        if response.result == GetLiveViewAuthResponse.SUCCESS:
+            message = json.dumps(
+                {
+                    "op": "authentication_response",
+                    "username": response.username,
+                    "msg": "",
+                    "permissions": response.permissions,
+                }
+            )
+            self.is_authenticated = True
+            self.username = response.username
+
+            for p in response.permissions:
+                parsed_permission = parse_permission(p)
+                if parsed_permission is not None:
+                    self.permissions[parsed_permission[0]].add(parsed_permission[1])
+
+            # Format a pleasant output of all permissions for user to console/logs.
+            permissions_string = ""
+            for action_name, resources in self.permissions.items():
+                permissions_string += f"\n{action_name}"
+                permissions_string += "".join(sorted([f"\n - {r}" for r in resources]))
+            rospy.loginfo(f"Authenticated user: {response.username} with permissions:{permissions_string}")
+
+        self.protocol.outgoing(message)
+
+    def has_permission(self, op: str, resourcePath: str) -> bool:
+        if op == "unsubscribe":
+            return True
+
+        # Walk all permissions for that operation to find a match. We use `endswith` because some resources might
+        # begin with a robot id.  eg.  `/p3_123`,  `/r2_12345`, `v1000`. There is no well-defined schema we can
+        # rely on, so we just compare if most/all of the remaining string is a known permission.
+        # Note that this is of limited security risk given we control both sides of this.
+        for permission in self.permissions[op]:
+            if resourcePath.endswith(permission):
```

Comment:
Just to confirm, this won't run into the same issue as service calls in FA (where `pause` and `unpause` were treated the same) because of the leading '/'.

Reply:
It will.  I was hoping to address it later once it all generally works. But if you've got a really simple solution, we can fix it now.

I'm trying to avoid getting into wacky regex  stuff because last time I tackled this I think my conclusion was that there's a ton of ambiguous edge cases.  I might be wrong.

Reply:
I don't, I'm fine tackling this later.